File: python_script/mc_combine.py
#    DataViewer visualization package
#    Copyright (C) 1997 Randy Paffenroth and Thomas Stone
#
#    This library is free software; you can redistribute it and/or
#    modify it under the terms of the GNU Library General Public
#    License as published by the Free Software Foundation; either
#    version 2 of the License, or (at your option) any later version.
#
#    This library is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#    Library General Public License for more details.
#
#    You should have received a copy of the GNU Library General Public
#    License along with this library; if not, write to the Free
#    Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
#    MA 02111-1307, USA
import tkFileDialog
import Tkinter
import Pmw
import string
import time

from geom_window import GeomWindow

def get_font():
    return "6x13"

def makeComboBox(parent,test_str,var):
    combo_box=Pmw.ComboBox(parent,labelpos="w",label_text=test_str,
                           selectioncommand = lambda val,obj=var:obj.set(val),
                           entryfield_entry_width = 5)
    combo_box._entryWidget.config(font=get_font())
    combo_box._list.component('listbox').config(font=get_font())
    combo_box.component('label').config(font=get_font())
    combo_box.component('entryfield').setentry(var.get())
    combo_box.pack()
    return combo_box

class MCGraph(GeomWindow):
    def __init__(self,parent=None,cnf={},**kw):
	apply(GeomWindow.__init__,(self,parent),kw)
	self.geom.set_problem("mc_combine")

    def UserWidgets(self):

	#file selection dialog
	self.filename = Tkinter.StringVar()
	self.filename.set("Problems/DrawGraph/dg10dv.txt")
        button = Tkinter.Button(self.UserTop,text="Select file to open....",
                                command=lambda obj=self:obj.Draw(tkFileDialog.askopenfilename(filetypes=[("graph files", "*.txt"),("all files", "*")])))
        button.pack()
        #redraw with current filename
        button = Tkinter.Button(self.UserTop,text="Redraw",
                                command=lambda obj=self:obj.Draw(obj.filename.get()))
        button.pack()

        some_frame=Tkinter.Frame(self.UserTop)
        some_frame.pack()
        
	self.sphere_factor = Tkinter.DoubleVar()
        self.sphere_factor.set(0.1667)
        line_frame=Tkinter.Frame(self.UserTop)
        line_frame.pack()
        line_typein = Pmw.ComboBox(line_frame,entry_textvariable=self.sphere_factor)
        line_typein.pack(side="right")
        line_label=Tkinter.Label(line_frame,text="Sphere radius factor")
        line_label.pack(side="left")

	self.epsilon_bf = Tkinter.DoubleVar()
	self.epsilon_bf.set(0.1)
        eps_box=makeComboBox(self.UserTop,"Epsilon bf",
                              self.epsilon_bf)
        eps_box.component('listbox').insert(1,0.05,0.01,0.1,0.2,0.5,0.9,1)

        self.iterations_bf=Tkinter.IntVar()
        self.iterations_bf.set(10)
        iter_box=makeComboBox(self.UserTop,"Number of iterations for bf",
                              self.iterations_bf)
        iter_box.component('listbox').insert(1,100,500,1000,5000,10000)

	self.epsilon_tv = Tkinter.DoubleVar()
	self.epsilon_tv.set(0.1)
        eps_box=makeComboBox(self.UserTop,"Epsilon tv",
                              self.epsilon_tv)
        eps_box.component('listbox').insert(1,0.05,0.01,0.005,0.001,0.1,0.2,0.5,0.9,1)

        self.iterations_tv =Tkinter.IntVar()
        self.iterations_tv.set(10)
        iter_box=makeComboBox(self.UserTop,"Number of iterations for tv",
                              self.iterations_tv)
        iter_box.component('listbox').insert(10,50,100,500,1000,5000,10000)
        
        self.step=Tkinter.IntVar()
        self.step.set(1)
        step_box=makeComboBox(self.UserTop,"Animation step",
                              self.step)
        step_box.component('listbox').insert(1,1,5,10,25,50,100)
        
	#Control widget for the geometry complexity
	self.geom_comp = Tkinter.IntVar()
	self.geom_comp.set(7)
	slider1 = Tkinter.Scale(some_frame,width=15,length=200)
	slider1["from"] = 1
	slider1["to"] = 9
	slider1["variable"] = self.geom_comp
	slider1["label"] = "Geom Complexity"
	slider1["orient"] = "horizontal"
	slider1["tickinterval"] = 1
	slider1["relief"] = "raised"
	slider1.pack(side="left",padx="2m")

	#Control widget for the light complexity
	self.light_comp = Tkinter.IntVar()
	self.light_comp.set(7)
	slider2 = Tkinter.Scale(some_frame,width=15,length=200)
	slider2["from"] = 1
	slider2["to"] = 9
	slider2["variable"] = self.light_comp
	slider2["label"] = "Light Complexity"
	slider2["orient"] = "horizontal"
	slider2["tickinterval"] = 1
	slider2.pack(side="left")
	slider2["relief"] = "raised"
	slider2.pack(side="left",padx="2m")


    def update_data(self,event):
	self.data = self.entry_contents.get()
    
    def Draw(self,filename=None):
        if filename:
            time1=time.time()
            fn=string.split(filename,'/')
            file_string=fn[len(fn)-1]
            print file_string
            self.UserTop.title(file_string)
            self.filename.set(filename)
            self.data = "%s %f %d %f %d %d %d %d %f" %(filename,
                                                       self.epsilon_bf.get(),
                                                       self.iterations_bf.get(),
                                                       self.epsilon_tv.get(),
                                                       self.iterations_tv.get(),
                                                       self.geom_comp.get(),
                                                       self.light_comp.get(),
                                                       self.step.get(),
                                                       self.sphere_factor.get())
            self.geom.set_problem("mc_combine")
            self.geom.file(self.data)
            print "Total time:", time.time() - time1
        elif self.filename.get():
            self.Draw(self.filename.get())
        

if __name__ == '__main__' : MCGraph().mainloop()

File: src/mc_combine.cc
/*
    DataViewer visualization package
    Copyright (C) 1997 Randy Paffenroth and Thomas Stone

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Library General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in thepy hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Library General Public License for more details.

    You should have received a copy of the GNU Library General Public
    License along with this library; if not, write to the Free
    Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
    MA 02111-1307, USA
*/
/*
  A library created by Dana Vrajitoru for the graph layout problem.  
*/

#include <iostream> 
#include <fstream> 
#include <strstream>
using namespace std;
#include <stdio.h>
#include <stdlib.h>
#include <cstring>
#include <ctype.h>
#include "common.h"

#include "DV.h"

#include "DrawGraph.h"
#include "GraphPos.h"

extern DVtips_and_hints global_th;
extern ofstream *fout;

/* Main function */

int mc_combine(DVcreate_image_arguments &dv)
{ 
  int iterations_bf, iterations_tv, step, frames1, frames2, intern = 1,
      geometry_complexity, light_complexity, i;  
  Graph *the_graph = new Graph;
  Point *points = NULL;
  float epsilon_bf, epsilon_tv, minw, maxw, last_norm = 0, sphere_factor;
  char filename[80], outfile[80];
  //  FILE *testfile;
  istrstream top_fin((char const*)dv.input_string,dv.input_string.getLength());
  
  top_fin >> filename ;
  top_fin >> epsilon_bf >> iterations_bf >> epsilon_tv >> iterations_tv;
  top_fin >> geometry_complexity;
  top_fin >> light_complexity; 
  top_fin >> step >> sphere_factor;

  frames1 = iterations_bf/step;
  frames2 = iterations_tv/step;
  cout << "frames1: " << frames1 << " frames2: " << frames2 << endl;

  dv.top.geometry_complexity = geometry_complexity;
  dv.top.light_complexity = light_complexity;
  dv.top.color[0] = DVcolor(1.0, 1.0, 1.0);
  DVselection_container cont;

  cout << "Reading graph from file: " << filename << endl;

  make_mc_out_name(filename, outfile, 3);
  cout << "Results into file:" << endl << outfile << endl;
  my_copy_file((char *)outfile);
  if (!fout)
    cout << "Could not open the file: " << outfile << endl;
  else {
  }

  the_graph->Read(filename);
  //cout << the_graph->weighted<< endl;

  //the_graph->Print();
  minw = the_graph->MinWeight();
  maxw = the_graph->MaxWeight();
  if (minw == 0)
    minw = maxw / 20;

  DVcolor_by_data tmp_color_by_data;
  DVcolor red(1, 0, 0), yellow(1, 1, 0), blue(0, 0, 1);
  DVcolor_map the_map(red, yellow, blue);
  tmp_color_by_data.set_color_map(the_map);
  tmp_color_by_data.set_min(-maxw);
  tmp_color_by_data.set_max(maxw);

  dv.top.color_by_data[0] = tmp_color_by_data;

  out_int("iterations_bf", iterations_bf);
  out_str("epsilon_bf: ");
  out_float(epsilon_bf);
  out_int("iterations_tv", iterations_tv);
  out_str("epsilon_tv: ");
  out_float(epsilon_tv);
  out_endl();
  out_int("vertices", the_graph->vertexNr);
  out_int("edges", the_graph->edgeNr);

  points = RandomPoints(the_graph->vertexNr, 1.5*maxw);
  out_float(the_graph->EvalDistance(points));
  if (the_graph->edgeNr <= 70)
    make_graph(cont, the_graph, points, minw*sphere_factor, 2);
  else
    make_graph(cont, the_graph, points, minw*sphere_factor, 1);
  out_endl();

  for (i=0; i<frames1; i++) {
    CreatePoints(points, the_graph, intern, step, epsilon_bf);

    out_float(the_graph->EvalDistance(points));
    if (i % 10 == 9)
      out_endl();

    if (the_graph->edgeNr <= 70)
      make_graph(cont, the_graph, points, minw*sphere_factor, 2);
    else
      make_graph(cont, the_graph, points, minw*sphere_factor, 1);
  }
  out_endl();
  out_endl();

  for (i=0; i<frames2; i++) {
    last_norm = VectorCreatePoints(points, the_graph, step, epsilon_tv);

    out_float(the_graph->EvalDistance(points));
    if (i % 10 == 9)
      out_endl();

    if (the_graph->edgeNr <= 70)
      make_graph(cont, the_graph, points, minw*sphere_factor, 2);
    else
      make_graph(cont, the_graph, points, minw*sphere_factor, 1);
  }
  out_endl();
  out_float(last_norm);
  out_endl();
  fout->close();

  dv.dvhints.set_first_frame_number(0);
  dv.dvhints.set_last_frame_number(frames1+frames2);
  dv.dvhints.set_animation_optimize_flag(1);
  dv.dvhints.set_animation_first_frame_number(0);
  dv.dvhints.set_animation_last_frame_number(frames1+frames2);
  dv.dvhints.set_animation_step_frame_number(1);
  dv.top.copy_child(cont);

  return 1;
}
